### PR TITLE
Support for sourcemaps

### PR DESCRIPTION
**Pull request by noctarius**
Support to read internal and external (but local) sourcemaps. Remote sourcemaps based on http/https schemes are ignored on purpose.

Errors / exception lines and columns are mapped via the loaded sourcemap and represent the real error location. This is specifically useful when running transpiled (e.g. Typescript) code.

### CODE REVIEW

File: parser/statement.go [hunk 1]
```diff
@@ -544,9 +549,48 @@ func (self *_parser) parseProgram() *ast.Program {
 		Body:            self.parseSourceElements(),
 		DeclarationList: self.scope.declarationList,
 		File:            self.file,
+		SourceMap:       self.parseSourceMap(),
 	}
 }
 
+func (self *_parser) parseSourceMap() *sourcemap.Consumer {
+	lines := strings.Split(self.str, "\n")
+	lastLine := lines[len(lines)-1]
```
Comment (dop251) [inline parser/statement.go]: I think this would be better:
```go
lastLine := self.str[strings.LastIndexByte(self.str, '\n') + 1:]
```

File: runtime.go [hunk 2]
```diff
@@ -119,7 +119,7 @@ func (f *stackFrame) write(b *bytes.Buffer) {
 	if f.prg != nil {
 		if n := f.prg.funcName; n != "" {
 			b.WriteString(n)
-			b.WriteString(" (")
+			b.WriteString(" ")
```
Comment (dop251) [inline runtime.go]: This is an unrelated change and should not really be part of this PR.

File: srcfile.go [hunk 3]
```diff
@@ -37,6 +40,19 @@ func (f *SrcFile) Position(offset int) Position {
 	if line >= 0 {
 		lineStart = f.lineOffsets[line]
 	}
+
+	row := line + 2
+	col := offset - lineStart + 1
+
+	if f.sourceMap != nil {
+		if _, _, row, col, ok := f.sourceMap.Source(row, col); ok {
```
Comment (dop251) [inline srcfile.go]: What about file name and function name?

File: srcfile.go [hunk 4]
```diff
@@ -37,6 +40,19 @@ func (f *SrcFile) Position(offset int) Position {
...
+			return Position{
+				Line: row,
+				Col:  col,
+			}
+		}
+	}
+
 	return Position{
 		Line: line + 2,
 		Col:  offset - lineStart + 1,
```
Comment (dop251) [inline srcfile.go]: The values are already in row and col variables, no need to calculate them again

File: runtime.go (hunk 2 above)
Comment (noctarius) [reply runtime.go]: I don't really think this is unrelated, as real lines/column number are a based requirement for meaningful debugging in THE golang ide right now (which obviously is Goland ;-)), anyhow gonna remove it.

File: srcfile.go (hunk 3 above)
Comment (noctarius) [reply srcfile.go]: Not sure where that would be exchanged with the sourcemap values. For me (since I load all the files myself) the values are already correct (the typescript compiler does not change method names).

Comment (dop251) [reply srcfile.go]: It will require a bit of refactoring... I'm going to merge this and get to it a bit later.

Comment (noctarius) [reply srcfile.go]: 👍 sure :)
